metrics/prometheus/prometheus.go

```diff
@@ -1,196 +1,202 @@
-				Name: &name,
-				Type: dto.MetricType_GAUGE.Enum(),
-				Metric: []*dto.Metric{{
-					Gauge: &dto.Gauge{
-						Value: &valFloat,
-					},
-				}},
-			})
-		case metrics.GaugeFloat64:
-			val := m.Snapshot().Value()
-			mfs = append(mfs, &dto.MetricFamily{
-				Name: &name,
-				Type: dto.MetricType_GAUGE.Enum(),
-				Metric: []*dto.Metric{{
-					Gauge: &dto.Gauge{
-						Value: &val,
-					},
-				}},
-			})
-		case metrics.Histogram:
-			snapshot := m.Snapshot()
-			count := snapshot.Count()
-			countUint := uint64(count)
-			sum := snapshot.Sum()
-			sumFloat := float64(sum)
-
-			ps := snapshot.Percentiles(pv)
-			qs := make([]*dto.Quantile, len(pv))
-			for i := range ps {
-				v := pv[i]
-				s := ps[i]
-				qs[i] = &dto.Quantile{
-					Quantile: &v,
-					Value:    &s,
-				}
+		mf, err := metricFamily(g.registry, name)
+		if err != nil {
+			if errors.Is(err, errMetricSkip) {
+				continue
 			}
+			return nil, err
+		}
+		mfs = append(mfs, mf)
+	}
 
-			mfs = append(mfs, &dto.MetricFamily{
-				Name: &name,
-				Type: dto.MetricType_SUMMARY.Enum(),
-				Metric: []*dto.Metric{{
-					Summary: &dto.Summary{
-						SampleCount: &countUint,
-						SampleSum:   &sumFloat,
-						Quantile:    qs,
-					},
-				}},
-			})
-		case metrics.Meter:
-			val := m.Snapshot().Count()
-			valFloat := float64(val)
-			mfs = append(mfs, &dto.MetricFamily{
-				Name: &name,
-				Type: dto.MetricType_GAUGE.Enum(),
-				Metric: []*dto.Metric{{
-					Gauge: &dto.Gauge{
-						Value: &valFloat,
-					},
-				}},
-			})
-		case metrics.Timer:
-			snapshot := m.Snapshot()
-			count := snapshot.Count()
-			countUint := uint64(count)
-			sum := snapshot.Sum()
-			sumFloat := float64(sum)
-
-			ps := snapshot.Percentiles(pv)
-			qs := make([]*dto.Quantile, len(pv))
-			for i := range ps {
-				v := pv[i]
-				s := ps[i]
-				qs[i] = &dto.Quantile{
-					Quantile: &v,
-					Value:    &s,
-				}
-			}
+	return mfs, nil
+}
 
-			mfs = append(mfs, &dto.MetricFamily{
-				Name: &name,
-				Type: dto.MetricType_SUMMARY.Enum(),
-				Metric: []*dto.Metric{{
-					Summary: &dto.Summary{
-						SampleCount: &countUint,
-						SampleSum:   &sumFloat,
-						Quantile:    qs,
-					},
-				}},
-			})
-		case metrics.ResettingTimer:
-			snapshot := m.Snapshot()
-
-			count := uint64(snapshot.Count())
-			if count == 0 {
-				continue
+var (
+	errMetricSkip = errors.New("metric skipped")
+)
+
+func ptrTo[T any](x T) *T { return &x }
+
+func metricFamily(registry Registry, name string) (mf *dto.MetricFamily, err error) {
+	metric := registry.Get(name)
+	name = strings.ReplaceAll(name, "/", "_")
+
+	switch m := metric.(type) {
+	case metrics.NilCounter, metrics.NilCounterFloat64, metrics.NilEWMA,
+		metrics.NilGauge, metrics.NilGaugeFloat64, metrics.NilGaugeInfo,
+		metrics.NilHealthcheck, metrics.NilHistogram, metrics.NilMeter,
+		metrics.NilResettingTimer, metrics.NilSample, metrics.NilTimer:
+		return nil, fmt.Errorf("%w: %q metric is nil", errMetricSkip, name)
```

Comment:
if we're going to skip this, can't we just return nil err?

Reply:
no we need to signal somehow to the caller to not collect that metric and to skip it